File: database-service/src/main/java/com/bnorm/barkeep/db/DbIngredientService.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.db;

import java.util.List;
import java.util.Optional;

import javax.persistence.EntityManager;

import com.bnorm.barkeep.model.Ingredient;
import com.bnorm.barkeep.service.IngredientService;


public class DbIngredientService extends AbstractDbService implements IngredientService {

  public DbIngredientService(EntityManager entityManager) {
    super(entityManager);
  }

  private IngredientEntity parentEntity(Ingredient ingredient) {
    Optional<Long> id = Optional.ofNullable(ingredient).map(Ingredient::getParent).map(Ingredient::getId);
    if (id.isPresent()) {
      IngredientEntity parent = findIngredient(id.get());
      // todo check parent is not null
      return parent;
    } else {
      return null;
    }
  }

  @Override
  public List<Ingredient> getIngredients() {
    return super.getIngredients();
  }

  @Override
  public IngredientEntity getIngredient(long ingredientId) {
    return super.getIngredient(ingredientId);
  }

  @Override
  public IngredientEntity createIngredient(Ingredient ingredient) {
    if (ingredient.getId() != null) {
      throw new IllegalArgumentException(String.format("Cannot create ingredient that already has an id=%d",
                                                       ingredient.getId()));
    }
    IngredientEntity parentEntity = parentEntity(ingredient);
    // todo check parentEntity exists if needed

    return transaction(em -> {
      IngredientEntity ingredientEntity = new IngredientEntity();

      ingredientEntity.setTitle(ingredient.getTitle());
      ingredientEntity.setParent(parentEntity);

      em.persist(ingredientEntity);
      return ingredientEntity;
    });
  }

  @Override
  public IngredientEntity setIngredient(long ingredientId, Ingredient ingredient) {
    IngredientEntity ingredientEntity = findIngredient(ingredientId);
    requireExists(ingredientEntity, ingredientId, "ingredient");
    requireMatch(ingredient, ingredientId, "ingredient");

    return transaction(em -> {
      if (ingredient.getTitle() != null) {
        ingredientEntity.setTitle(ingredient.getTitle());
      }
      if (ingredient.getParent() != null) {
        ingredientEntity.setParent(parentEntity(ingredient));
      }

      return ingredientEntity;
    });
  }

  @Override
  public void deleteIngredient(long ingredientId) {
    transaction(em -> {
      IngredientEntity ingredientEntity = findIngredient(ingredientId);
      requireExists(ingredientEntity, ingredientId, "ingredient");
      em.remove(ingredientEntity);
    });
  }
}

File: service-database/src/main/kotlin/com/bnorm/barkeep/db/DbIngredientService.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.db

import com.bnorm.barkeep.model.Ingredient
import com.bnorm.barkeep.model.IngredientSpec
import com.bnorm.barkeep.service.IngredientService
import javax.persistence.EntityManager

class DbIngredientService(entityManager: EntityManager) : AbstractDbService(entityManager), IngredientService {

  private fun parentEntity(ingredient: IngredientSpec): IngredientEntity? {
    return ingredient.parent?.id.let { findIngredient(it) }
  }

  override fun getIngredients(): List<Ingredient> {
    return super.getIngredients()
  }

  override fun getIngredient(id: Long): IngredientEntity? {
    return super.getIngredient(id)
  }

  override fun createIngredient(ingredient: IngredientSpec): IngredientEntity {
    val parentEntity = parentEntity(ingredient)
    // todo check parentEntity exists if needed

    val ingredientEntity = IngredientEntity()

    ingredientEntity.title = ingredient.title
    ingredientEntity.parent = parentEntity

    txn {
      em.persist(ingredientEntity)
    }

    return ingredientEntity
  }

  override fun setIngredient(id: Long, ingredient: IngredientSpec): IngredientEntity {
    val ingredientEntity = requireExists(findIngredient(id), id, "ingredient")

    txn {
      if (ingredient.title != null) {
        ingredientEntity.title = ingredient.title
      }
      if (ingredient.parent != null) {
        ingredientEntity.parent = parentEntity(ingredient)
      }
    }

    return ingredientEntity
  }

  override fun deleteIngredient(id: Long) {
    txn {
      val ingredientEntity = findIngredient(id)
      requireExists(ingredientEntity, id, "ingredient")
      em.remove(ingredientEntity)
    }
  }
}
